File: src/main/java/references/streams_exception_handling/ak_hanlding_exception_in_the_pipeine_of_CF/Sample.java
package references.streams_exception_handling.ak_hanlding_exception_in_the_pipeine_of_CF;

import java.net.URL;
import java.util.List;
import java.util.Scanner;
import java.util.concurrent.CompletableFuture;

public class Sample {
    public static String getNameOfAirport(String iata) {
        try {
            var url = "https://soa.smext.faa.gov/asws/api/airport/status/" + iata;

            try (var scanner = new Scanner(new URL(url).openStream())) {
                var response = scanner.nextLine();

                if (!response.contains("Name")) {
                    throw new RuntimeException("Invalid airport code " + iata);
                }

                return response.split("\"")[3];
                //way too lazy to do the real work to get the data
            }
        } catch (Exception ex) {
            throw new RuntimeException(ex);
        }
    }

    public static CompletableFuture<Void> processIATA(String iataCode) {
        return CompletableFuture.supplyAsync(() -> getNameOfAirport(iataCode))
                .thenApply(String::toUpperCase)
                .exceptionally(Throwable::getMessage)
                .thenAccept(System.out::println);
    }

    public static void main(String[] args) {
        var iataCodes = List.of("AUS", "IAH", "DFW", "TAS", "SAT");

        var tasks = iataCodes.stream()
                .map(Sample::processIATA)
                .toList();

        try {
            Thread.sleep(10000);
        } catch (Exception _) {
        }
    }
}

//A bit better with a good separation of concern for each function
//then... focuses on data
//exceptionally focuses on error
//
//But, we still have to deal with a mixture of data handling and error
//handling in different stages of the pipeline.

File: src/main/java/references/streams_exception_handling/scala/03_using_try/sample.scala
import scala.util.{ Try, Success, Failure }

def getNameOfAirport(iata: String) = {
  val url = s"https://soa.smext.faa.gov/asws/api/airport/status/$iata"
  
  val response = scala.io.Source.fromURL(url).mkString 

  if(!response.contains("Name")) {
    throw new RuntimeException("Invalid airport code " + iata)
  }

  response.split("\"")(3)
  //way too lazy to do the real work to get the data
}

def tryToGetAirportName(iata: String) = Try(getNameOfAirport(iata))

val iataCodes = List("AUS", "IAH", "DFW", "TAS", "SAT")

iataCodes
  .map { tryToGetAirportName }
  .map { _.map { _.toUpperCase } }
  .map { _ match {
      case Success(name) => name
      case Failure(ex) => ex.getMessage
    }
  }
  .foreach(println)
